stdlib/public/core/EmbeddedRuntime.swift

```diff
@@ -291,6 +291,10 @@ func swift_release_n_(object: UnsafeMutablePointer<HeapObject>?, n: UInt32) {
 
   let resultingRefcount = subFetchAcquireRelease(refcount, n: Int(n)) & HeapObject.refcountMask
   if resultingRefcount == 0 {
+    // Set the refcount to immortalRefCount before calling the object destroyer
+    // to prevent future retains/releases from having any effect.
+    storeRelaxed(refcount, newValue: HeapObject.immortalRefCount)
```

Comment:
Maybe worth commenting explicitly that there can only be one thread holding a reference to this object at this point (unless there's a bug somewhere), which is why a relaxed store is fine here.

Reply:
Seconded. On another note, the non-embedded runtime continues tracking refcounts, and this allows it to raise a fatal error when `self` escapes from `deinit` by checking the refcount when the memory is freed. See here:

https://github.com/swiftlang/swift/blob/03abb1f0f84e9e3f57083f246cf9ba648b2d5c12/stdlib/public/runtime/HeapObject.cpp#L850

Obviously this is an optional feature and not something you need to do now (or ever), but might be worth keeping in mind as a possible enhancement.